### ISSUE TITLE
direct-read-namespaces for 1.1 billion entries

### ISSUE THREAD

**Issue by ChessSpider** [rwynn/monstache#190]
Hi, I have a mongodb db w/ 53 collections for a total of 1,1 billion documents. 

I need to insert these into elasticsearch, and I was hoping for some guidance on how to do this as quickly as possible. 

Both run, dockerized, on a bare-metal i7 with 24gigs of RAM. I set the mongodb `wiredTigerCacheSizeGB`  to 10. I limited ELK-stack to 8g of RAM `"ES_JAVA_OPTS=-Xms8g -Xmx8g"`.  Additionally, queue_size is set to 200.  CPU load averages are at `18.90 20.79 22.88`. Memory is at 20/23gb used. Unfortunately both the mongodb and elk stack are on the same two ZVOL'ed SSDs. 

The logfile of monstache is still spammed w/
```
ERROR 2019/03/22 14:56:44 Bulk response item: {"_index":"importpw.qm","_type":"_doc","_id":"5c8d630213ab3a58ec2d056d","status":429,"error":{"type":"es_rejected_execution_exception","reason":"rejected execution of processing of [1386143][indices:data/write/bulk[s][p]]: request: BulkShardRequest [[importpw.qm][0]] containing [660] requests, target allocation id: MBIIfJ5aSOaCh1u4RiiTkQ, primary term: 1 on EsThreadPoolExecutor[name = fAhQBq0/write, queue capacity = 200, org.elasticsearch.common.util.concurrent.EsThreadPoolExecutor@4f2fe7b6[Running, pool size = 8, active threads = 8, queued tasks = 200, completed tasks = 486895]]"}}
```

The indexing is still set to 1s, so I know that doesnt help, but are there more things I can do?

Monstache settings are simply:
```
    environment:
      - MONSTACHE_MONGO_URL=127.0.0.1:27017
      - MONSTACHE_ES_URL=127.0.0.1:9200
      - MONSTACHE_DIRECT_READ_NS=fiftythree,collections,comma,separated,are,here
```

**Comment by rwynn** [comment]
Couple of things you can try.  

- Reduce `direct-read-split-max` to 1.  Each namespace (53 in your case) is split by default into 10 segments via the default split count of 9.  Each segment is read concurrently in a separate go routine.  By using a value of 1 you will have 2 segments instead of 10 and much less go routines in total across all namespaces.
- Use less concurrent connections to Elasticsearch via `elasticsearch-max-conns` (defaults to 4) or decrease the `elasticsearch-max-bytes` defaults to 8000000 (8mb size bulk requests).  For each connection the buffer will fill to, by default, 8MB, before sending the request. Elasticsearch may be better at handling less concurrency or smaller payloads.  
- Set the replica count to 0 while you are doing the indexing.  This will make things easier for Elasticsearch.  After the indexing is complete you can up the replica count on the indexes.  
- Increase the value of `indices.memory.index_buffer_size` for Elasticsearch.  This gives more memory for indexing and the default is 10% of available RAM.  

You can find examples of the last 2 in https://github.com/rwynn/monstache-showcase.  See elasticsearch.yml and template.json.

**Comment by rwynn** [comment]
If none of that helps and you cannot scale Elasticsearch horizontally (I'm assuming you need everything on the same machine), you might try using docker's `cpus/cpu_count` variable to limit the number of cpus that the monstache container can use.  This will free the cpu cycles for indexing and slow the speed at which documents are sent for indexing.  

Most docs I see on [429s](https://www.elastic.co/blog/why-am-i-seeing-bulk-rejections-in-my-elasticsearch-cluster) seem to recommend increasing the number of nodes in the Elasticsearch cluster.

**Comment by rwynn** [comment]
Also, I think it's recommended to ensure that Elasticsearch is memory locked and not using swap if not already:

```yaml
environment:
      - bootstrap.memory_lock=true
      - "ES_JAVA_OPTS=-Xms8g -Xmx8g"
    ulimits:
      memlock:
        soft: -1
        hard: -1
```

**Comment by ChessSpider** [comment]
Thank you very much. I just stopped monstache after disabling `refresh_interval`. Es was using a lot of performance in the first few minutes, now its calm. Restarted monstache, and not a lot seem to be happening now. Confusing

**Comment by ChessSpider** [comment]
Wait, is it trying to  do sync all indexes in serial or parallel ? If the latter, that's not gonna work. It has to do them one by one, due to RAM

**Comment by rwynn** [comment]
It does them all in parallel.  I think maybe for next version I will support an option to index direct reads serially.  For now you might just want to run it in smaller batches.

**Comment by rwynn** [comment]
In your case with the defaults it would be 53 times potentially 10 go routines in parallel.  The split count may be less for smaller collections.  The logic for the direct reads and splitting starts [here](https://github.com/rwynn/gtm/blob/master/gtm.go#L1722).

**Comment by ChessSpider** [comment]
hm yeah, total index size for all colls in mongodb is like 85gb so that'snot great haha.
I now started it w/ 2 indexes, but its difficult to know if it's effectively using all resources now. CPU usage goes between 20% and 100%, a lot around 70%.

**Comment by rwynn** [comment]
I should be more clear though.  Monstache will not endlessly eat memory.  If you look at the rwynn/gtm code the channels have a buffer size.  This defaults to 512.  Monstache will not read more from MongoDB until there is room on that channel to send (otherwise it will block). That channel only gets emptied when the indexing go routines are free to read off a document to index it.   So what I wanted to say is that there is back pressure across monstache due to how go channels work.

**Comment by ChessSpider** [comment]
Yeah but mongodb will try to fit the index into ram for searching is my
experience. If all 53 collections are searched, it's a problem.

**Comment by rwynn** [comment]
Ok, I see what you mean.  Yes, it will have multiple cursors open to all those collections at the same time.

**Comment by rwynn** [comment]
When I run that previously mentioned `monstache-showcase` docker compose project in a single VM with 14GB I get this relative distribution.

```
PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
4315 vagrant   20   0 9540684 5.774g 187420 S 148.2 43.1   6:44.31 java                                                                                     
 4658 root      20   0 3440276 1.652g  34440 S  65.4 12.3   2:17.91 mongod                                                                                   
 5320 root      20   0  964836 845440  17004 S  61.1  6.0   5:11.68 monstache
```
What does it look like on your end?

At least in that case it seems java (elasticsearch) has the lion's share of resources.  I don't get any errors in this case.  Maybe try to configure to push resources to Elasticsearch?  

Of course this is a single mongodb collection with only 6.5 million records.  Not sure if there is a way to limit MongoDB's available resources?  The original problem was Elasticsearch not able to keep up hence the 429 so just a thought.

**Comment by ChessSpider** [comment]
6.5m documents? Oh that's cute ;) ;)  j/k

This is what it looks like on my machine trying to sync all 53 colls: https://imgur.com/zkw6Za2 Note the amount of mongodb threads. I'd assume that by reading from all 53 colls at once, mongodb starves es out of io and cpu.

This is w/ `direct-read-split-max= 1` and `exit-after-direct-reads= true`. I noticed by the way that monstache doesn't do an 'exit 0' with exit-after-direct-reads set? That does make my life a bit harder with batching monstache.

**Comment by rwynn** [comment]
It exits with os.Exit(exitStatus). exitStatus only changes to 1 on any error encountered during the sync. It must have had some error if you’re not seeing 0.

**Comment by ChessSpider** [comment]
Weird, this is my docker-compose:
```
version: '2.4'
services:
  monstache:
    network_mode: host
    image: rwynn/monstache:latest
    ports:
      - "8080:8080"
    volumes:
      - ./config.toml:/config.toml:ro
    entrypoint: monstache -f /config.toml
    cpus: 2
    logging:
      driver: "json-file"
      options:
        max-size: "200k"
        max-file: "2"
```
this is my config.toml:
```
direct-read-namespaces = ["......", ".... ]
direct-read-split-max= 3
exit-after-direct-reads= true
```

Yet, docker-compose says it's still running and doesnt exit

bythe way, compliments on the speed. Ït imports 100m items within an hour, without even an half-assed attempt from my to optimize it. Thx!

**Comment by rwynn** [comment]
That's weird about it not exiting eventually.  I've tested that in the past.  Does -print-config report exit.. as true?  When the "Direct read completed" is printed then it should drain all buffers and eventually call exit with a timeout of 30s.  The timeout is just if it cannot cleanly shutdown the connections.  

By the way, I'm working on 2 things. Allow split-max to be 0 in which case only 1 cursor (go routine) is used for the entire collection and splitting is not attempted.  The other is to allow setting the direct read concurrency. It will be an int and only that amount of collections will be read from at a time.  Defaults to current behavior of reading all concurrently.

**Comment by rwynn** [comment]
By the way you can set `disable-change-events` to true if you are not interested in new MongoDB activity being synched.  This would make monstache not tail the oplog for new events.

**Comment by ChessSpider** [comment]
> 
> 
> By the way you can set `disable-change-events` to true if you are not interested in new MongoDB activity being synched. This would make monstache not tail the oplog for new events.

that did the trick Thanks!




> By the way, I'm working on 2 things. Allow split-max to be 0 in which case only 1 cursor (go routine) is used for the entire collection and splitting is not attempted. The other is to allow setting the direct read concurrency. It will be an int and only that amount of collections will be read from at a time. Defaults to current behavior of reading all concurrently.

😍

**Comment by rwynn** [comment]
@ChessSpider you can now use the latest release with `direct-read-concur` set to 1 to do the direct reads serially.  Or you can set it to a higher value to sync concurrently in batches.  It will still sync all concurrently if you don't override this option.

**Comment by rwynn** [comment]
Also, you can now set `direct-read-split-max` to -1 to disable collection splitting. That will give you only 1 cursor open per collection.

**Comment by ChessSpider** [comment]
Awesome! Testing it now.

**Comment by ChessSpider** [comment]
Seems to work great, thanks! 
It has been processing for the past ~7 hours now, and i have no idea how far it is now.  Is there any way to know which collections have finished?
```
monstache_1  | INFO 2019/03/27 12:58:17 Started monstache version 4.16.0
monstache_1  | INFO 2019/03/27 12:58:17 Successfully connected to MongoDB version 4.0.6
monstache_1  | INFO 2019/03/27 12:58:17 Successfully connected to Elasticsearch version 6.6.2
monstache_1  | INFO 2019/03/27 12:58:17 Listening for events
```

Also, but possibly another issue, as mentioned in https://github.com/rwynn/monstache/issues/190#issuecomment-475883111, I currently have both set:
```
exit-after-direct-reads = true
disable-change-events= true
```
with the latter seem to be required to actually make monstache exit(0) after direct-reads finished. Is the `exit-after-direct-reads` working as indended?

Anyway thank you very much now, this is great!

**Comment by rwynn** [comment]
There is not any way to know which collections have finished because I didn’t add a log statement for each. Only when all have completed it logs a message. The only thing you could do now is check the counts across MongoDB and Elasticsearch.

**Comment by rwynn** [comment]
You should not require change events to be disabled for exit-after... to work. I tested this yesterday and it worked ok in my case. These collections were much smaller than in your case however. 

When change events are enabled, stopping Monstache takes longer. One reason for this is because the cursor that tails the oplog is configured to wait for events instead of returning immediately when there are none. There is a timeout associated defaulting to 10s. If no new events occur in this time then a timeout occurs and Monstache checks to see if a certain go channel has been closed. If so it should return from the function and allow shutdown to proceed. Also, the function checks this channel after reading each event. This is to ensure that it can return even if a continuous stream of events are coming in. 

The code for all of that is in rwynn/gtm on github. The function that would be holding up shutdown is named TailOps here -> https://github.com/rwynn/gtm/blob/development/gtm.go

**Comment by ChessSpider** [comment]
weird, well whatever, its working now so that's great. Thank you very much for your support and quick development 👍
